### PR TITLE
NDS: merge name info by hostname

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
When a SE has multiple vips, we should merge those ips into the DNS name info instead of overriding
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: svc
spec:
  hosts:
  - foo.bar
  addresses:
  - 10.0.0.5
  - 10.0.0.6
  ports:
  - name: tcp
    number: 33333
    protocol: tcp
  location: MESH_EXTERNAL
```

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
Originally the addreesses in serviceentry are not to be resolved by DNS proxy. It is used to match traffic's dest IP.  Not sure if itmatters

**Comment by dddddai** [comment]
I think it does matter: https://istio.io/latest/docs/ops/configuration/traffic-management/dns-proxy/#dns-capture-in-action

**Comment by hzxuzhonghu** [comment]
Multi adrresses is not meant to be used for DNS resolving. It is to match dst addr.

**Comment by dddddai** [comment]
Regardless of multiple addresses, the same happens to SE as a "decorator" of k8s Service? If we don't merge, the DNS proxy would not be able to resolve short names

We are doing a similar check when building sidecar scopes
https://github.com/istio/istio/blob/1.16.2/pilot/pkg/model/sidecar.go#L200-L209

**Comment by dddddai** [comment]
Regardless of multiple addresses, the same happens to [SE as a "decorator" of k8s Service](https://github.com/istio/api/blob/1.16.2/networking/v1beta1/service_entry.proto#L824)? If we don't prioritize k8s Service, the DNS proxy would not be able to resolve short names

We are doing a similar check when building sidecar scopes
https://github.com/istio/istio/blob/1.16.2/pilot/pkg/model/sidecar.go#L200-L209

File: pkg/dns/server/name_table.go [hunk 1]
```diff
@@ -114,21 +114,37 @@ func BuildNameTable(cfg Config) *dnsProto.NameTable {
...
+				// to avoid conflicting with the entries for the regular (cluster.local) service.
+				nameInfo.Namespace = svc.Attributes.Namespace
+				nameInfo.Shortname = svc.Attributes.Name
+			}
+			out.Table[hostName.String()] = nameInfo
+		} else if provider.ID(ni.Registry) != provider.Kubernetes {
+			// 2 possible cases:
+			// 1. If the SE has multiple addresses(vips) specified, merge the ips
+			// 2. If the previous SE is a decorator of the k8s service, give precedence to the k8s service
+			if svc.Attributes.ServiceRegistry == provider.Kubernetes {
```
Comment (hzxuzhonghu) [inline pkg/dns/server/name_table.go:137]: check `!strings.HasSuffix(hostName.String(), "."+constants.DefaultClusterSetLocalDomain)`

File: pkg/dns/server/name_table_test.go [hunk 2]
```diff
@@ -168,6 +168,30 @@ func TestNameTable(t *testing.T) {
...
+			Namespace:       "testns",
+			ServiceRegistry: provider.External,
+		},
+	}
+	serviceWithVIP2 := serviceWithVIP1.DeepCopy()
+	serviceWithVIP2.DefaultAddress = "10.0.0.6"
+
+	decoratedService := serviceWithVIP1.DeepCopy()
+	decoratedService.DefaultAddress = "10.0.0.7"
+	decoratedService.Attributes.ServiceRegistry = provider.Kubernetes
```
Comment (hzxuzhonghu) [inline pkg/dns/server/name_table_test.go:193]: Add another case: k8s service created ealier than SE

File: pkg/dns/server/name_table.go (hunk 1 above)
Comment (dddddai) [reply pkg/dns/server/name_table.go:137]: Checked at L140

**Review by hzxuzhonghu — APPROVED** [review]

**Review by GregHanson — CHANGES_REQUESTED** [review]
since I don't think there is a linked issue for this one - could you link this PR (full url) in the `issues` field so the generated release notes link back to this.  This way users will be able to see your example ServiceEntry to determine whether or not this is something affecting them

**Review by GregHanson — APPROVED** [review]
release note looks good

**Pull request merged**